Send events before adding watchers in traversePluginDir

bertinatto opened the pull request:
**What type of PR is this?**

/kind bug

**What this PR does / why we need it**:

This PR changes the function `devicemanager.traversePluginDir` in order to send events before adding watchers to directories.

**Which issue(s) this PR fixes**:

Fixes #75097

**Does this PR introduce a user-facing change?**:

```release-note
NONE
```
/sig storage
/sig node
/assign @liggitt

liggitt commented on pkg/kubelet/util/pluginwatcher/plugin_watcher.go:
```diff
@@ -206,25 +211,36 @@ func (w *Watcher) traversePluginDir(dir string) error {
+			dirsToWatch = append(dirsToWatch, path)
 		case mode&os.ModeSocket != 0:
-			w.wg.Add(1)
-			go func() {
```
why accumulate rather than continuing to send async here? we can't assume `w.fsWatcher.Events` won't block.

bertinatto replied:
Yeah, I noticed that after later...

liggitt commented on pkg/kubelet/util/pluginwatcher/plugin_watcher.go:
```diff
@@ -206,25 +211,36 @@ func (w *Watcher) traversePluginDir(dir string) error {
 			if w.containsBlacklistedDir(path) {
 				return filepath.SkipDir
 			}
-
```
starting watching the dir after we descend and scan it for socket files means that we can miss new socket files that appear. we *have* to start watching before we scan the content.

bertinatto replied:
Yes, there'd be another race if the file were created after we descend and before start watching.

I changed the order of events to:

1. Register file system watchers as we descend
2. Enqueue and process socket files found
3. Start processing socket files observed by watchers added in step 1

bertinatto replied:
Done.

bertinatto committed:
Process plugins found before the ones watched

Currently, the method `pluginwatcher.traversePluginDir` descends into
a directory adding filesystem watchers and creating synthetic `create`
events when it finds sockets files. However, a race condition might
happen when a recently-added watcher observes a `delete` event in a
socket file before `pluginwatcher.traversePlugindir` itself notices
this file.

This patch changes this behavior by registering watchers on
directories, enqueueing and processing `create` events from sockets
found, and only then processing the events from the registered watchers.

bertinatto commented:
/retest

bertinatto commented:
/remove-wip

bertinatto commented:
/remove-work-in-progress

bertinatto commented:
/remove-label do-not-merge/work-in-progress

liggitt commented on pkg/kubelet/util/pluginwatcher/plugin_watcher.go:
```diff
@@ -191,6 +220,15 @@ func (w *Watcher) init() error {
 // Goroutines started here will be waited for in Stop() before cleaning up.
 // Ignore all errors except root dir not being walkable
 func (w *Watcher) traversePluginDir(dir string) error {
```
this is called at times other than Start()... checking for isPluginDir() doesn't ensure that scanned synthetic events from new directories we encounter are processed before observed events for those new directories

bertinatto replied:
Done

liggitt commented on pkg/kubelet/util/pluginwatcher/plugin_watcher.go:
```diff
@@ -191,6 +220,15 @@ func (w *Watcher) init() error {
 // Goroutines started here will be waited for in Stop() before cleaning up.
 // Ignore all errors except root dir not being walkable
 func (w *Watcher) traversePluginDir(dir string) error {
```
would it be simpler to:
1. have traversePluginDir call `w.handleCreateEvent` directly and synchronously, rather than enqueuing synthetic create events in the event channel
2. wait until the two traversePluginDir() calls have completed in Start() before starting the event processing goroutine

that would ensure that while traversing a given directory (either in Start() or in response to an observed directory creation), no observed events from the started watchers were processed

bertinatto replied:
Done

rphillips commented:
/lgtm

@liggitt it appears your comments were addressed. Can you give this PR another review?

bertinatto commented:
Also,

/assign @vishh 

for approval.

liggitt commented on pkg/kubelet/util/pluginwatcher/plugin_watcher.go:
```diff
@@ -214,10 +208,10 @@ func (w *Watcher) traversePluginDir(dir string) error {
 			w.wg.Add(1)
 			go func() {
```
don't we have to remove this goroutine, so that `traversePluginDir` runs synchronously and completes before we start our processing loop (and before we continue processing other events in the event loop)?

bertinatto replied:
Fixed. I thought starting the goroutine would be enough to guarantee that it'd be executed before the processing goroutine, but I was wrong.

liggitt commented on pkg/kubelet/util/pluginwatcher/plugin_watcher.go:
```diff
@@ -214,10 +208,10 @@ func (w *Watcher) traversePluginDir(dir string) error {
 			go func() {
 				defer w.wg.Done()
-				w.fsWatcher.Events <- fsnotify.Event{
+				w.handleCreateEvent(fsnotify.Event{
```
handle the returned error (log for now, but add a matching `//TODO: Handle errors by taking corrective measures`)

bertinatto replied:
Done.

bertinatto committed:
Remove unnecessary in-line function

bertinatto commented on pkg/kubelet/util/pluginwatcher/plugin_watcher_test.go:
```diff
@@ -189,6 +186,9 @@ func TestPluginRegistrationAtKubeletStart(t *testing.T) {
 
+	w := newWatcherWithHandler(t, hdlr, false /* testDeprecatedDir */)
+	defer func() { require.NoError(t, w.Stop()) }()
+
```
Since we're processing serially now, there're would be a lot of contention [here](https://github.com/bertinatto/kubernetes/blob/832822b41808f54f563fa513dbb3371164803e44/pkg/kubelet/util/pluginwatcher/example_plugin.go#L103). To avoid that, this starts the channel-reading goroutines earlier.

liggitt commented:
this lgtm and is easier to follow, in my opinion. will leave final ack to sig-node

bertinatto commented:
/test pull-kubernetes-bazel-test

bertinatto commented:
/retest

bertinatto commented:
/test pull-kubernetes-bazel-test

bertinatto commented:
/test pull-kubernetes-bazel-test

bertinatto commented:
/retest

Something is going when jobs try to `merge` this branch. These error are unrelated to the changes from this patch.

bertinatto commented:
/assign @kubernetes/sig-node-pr-reviews

liggitt commented:
/skip

liggitt commented:
all the required CI jobs are green on this PR, ready for node/storage review

bertinatto commented:
/wip cancel

bertinatto commented:
/remove-work-in-progress

bertinatto commented:
/remove-wip

bertinatto commented:
/remove-do-not-merge/work-in-progress

bertinatto commented:
/remove-label do-not-merge/work-in-progress

tallclair commented on pkg/kubelet/util/pluginwatcher/plugin_watcher.go:
```diff
@@ -211,14 +205,14 @@ func (w *Watcher) traversePluginDir(dir string) error {
+				Op:   fsnotify.Create,
+			}
+			//TODO: Handle errors by taking corrective measures
+			if err := w.handleCreateEvent(event); err != nil {
```
Looking at the implementation of handleCreateEvent, I think it's more direct to just do:
```go
if !w.containsBlacklistedDir(path) {
  return w.traversePluginDir(path)
}
```

bertinatto replied:
Not sure if I follow, do you mean inside `handleCreateEvent`? If so, then it wouldn't call `handlePluginRegistration`... or am I missing somehting?

tallclair replied:
I mean "handle create event" doesn't really seem like what this is doing, unless I'm not following the control flow. If you look at what handle create event is doing, the snippet I commented above is all that gets hit in that function, so I'm suggesting to just inline that piece here.

bertinatto replied:
Ah, OK... I understand now...

This part of the code creates "synthetic create events" if we find a socket file. Then we "handle" this event right here (as opposed to sending it to a processing goroutine like it was before).

Since we're dealing with a socket file, we  don't really want to `traversePluginDir()` it, but instead we want to register the plugin that's listening to this unix domain socket file (i.e., `handlePluginRegistration()`).

However, we also need to make sure that we ignore black-listed directories (like you pointed in the snippet above) and files prefixed with a "."; this would do it:

```go
			if !w.containsBlacklistedDir(path) {
				if !strings.HasPrefix(path, ".") {
					// TODO: Handle errors by taking corrective measures
					if err := w.handlePluginRegistration(path); err != nil {
						klog.Errorf("error %v when handling create event for file: %s", err, path)
					}
				}
			}
```

IMO `handleCreateEvent()` looks like would be a better fit though (since we're almost re-implementing it in-line).

Does this make sense?

tallclair replied:
Ah, sorry, I completely misread what `handleCreateEvent` was doing.

tallclair commented on pkg/kubelet/util/pluginwatcher/plugin_watcher_test.go:
```diff
@@ -238,11 +238,22 @@ func TestPlugiRegistrationFailureWithUnsupportedVersionAtKubeletStart(t *testing
+	select {
+	case <-c:
+		return
+	case <-time.After(2 * time.Second):
```
```suggestion
	case <-time.After(wait.ForeverTestTimeout):
```

You can't wait timeout before that in a unit test, as it will flake in the CI environment.

bertinatto replied:
Fixed. I took the chance and fixed the other `time.After` calls from this file.

tallclair commented on pkg/kubelet/util/pluginwatcher/plugin_watcher_test.go:
```diff
@@ -238,11 +238,22 @@ func TestPlugiRegistrationFailureWithUnsupportedVersionAtKubeletStart(t *testing
+		defer close(c)
+		require.True(t, waitForEvent(t, exampleEventValidate, hdlr.EventChan(p.pluginName)))
+		require.False(t, waitForPluginRegistrationStatus(t, p.registrationStatus))
+	}()
```
alternatively, you can just do this serially at the end of the test and rely on the whole test timing out for the failure condition.

bertinatto replied:
Since we're processing plugins serially now, `newWatcherWithHandler` below would block forever trying to write to the event/registration channels. This goroutine prevents that by receiving info from these channels beforehand.

bertinatto commented:
skip

bertinatto commented:
/skip

bertinatto commented:
/retest

bertinatto commented:
/skip

bertinatto commented:
/retest

bertinatto commented:
/skip

bertinatto commented:
/retest

bertinatto commented:
/retest

sjenning commented:
/assign @derekwaynecarr

tallclair commented:
/lgtm

tallclair commented:
/approve

tedyu commented on pkg/kubelet/util/pluginwatcher/plugin_watcher.go:
```diff
@@ -211,14 +205,14 @@ func (w *Watcher) traversePluginDir(dir string) error {
+				Op:   fsnotify.Create,
+			}
+			//TODO: Handle errors by taking corrective measures
+			if err := w.handleCreateEvent(event); err != nil {
```
Shouldn't the err be returned, in the same way error is returned for mode.IsDir() case ?
